Add dispatcher events for audit log

sharidas opened the pull request:
Add dispatcher events for audit log

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented:
Tests done are:
- [x] Create a new custom group. Verify the logs are generated.
- [x] Add users to the custom group. Verify the logs are generated.
- [x] Remove user(s) from the group. Verify the logs are generated.
- [x] Remove custom group. Verify the logs are generated.

PVince81 commented:
@sharidas from now on we should rather use the event dispatcher for hooks. I believe you've used it before ? You'll likely need to inject it.

PVince81 commented:
As for the class name in the hook name I suggest to just use `\OCA\CustomGroups` without detailed class name, because that would be too deep an implementation detail.

For the event name I think it would also be better to remove the "customgroup_" prefix as we're already anyway inside the `\OCA\CustomGroups` scope so no need for additional prefixing.

sharidas commented:
Updated the PR with dispatcher instead of hooks.

PVince81 commented on lib/CustomGroupEvent.php:
```diff
@@ -0,0 +1,44 @@
+
+class CustomGroupEvent extends GenericEvent {
+
+	protected $order;
```
order ? was is this ?

```
$order = "One burger please with a side of fries";
```

PVince81 replied:
(answer in PHPDoc comment)

PVince81 commented on lib/Dav/GroupsCollection.php:
```diff
@@ -100,6 +106,9 @@ public function createDirectory($name) {
 		// add current user as admin
 		$this->groupsHandler->addToGroup($this->helper->getUserId(), $groupId, true);
+
+		$this->event->addData(['groupName' => $name, 'user' => $this->helper->getUserId()]);
```
you could also simply use Symfony's GenericEvent which is usually enough to contain a hash.

Is there a use case for having more than one set of data ?

sharidas commented:
Improvised the PR, by removing unnecessary changes made by creating CustomGroupEvent.php. Now the change uses GenericEvent to handle the data used for logging.

PVince81 commented on lib/Dav/GroupsCollection.php:
```diff
@@ -100,6 +101,9 @@ public function createDirectory($name) {
 		$this->groupsHandler->addToGroup($this->helper->getUserId(), $groupId, true);
+
+		$event = new GenericEvent(null, ['groupName' => $name, 'user' => $this->helper->getUserId()]);
+		\OC::$server->getEventDispatcher()->dispatch('addGroupAndUser', $event);
```
looks good so far, next up: have the event dispatcher injected in the constructor

sharidas commented:
Dispatcher moved to constructor.

PVince81 commented on lib/Dav/GroupMembershipCollection.php:
```diff
@@ -77,6 +90,9 @@ public function __construct(
 		$this->groupInfo = $groupInfo;
 		$this->helper = $helper;
+
+		$this->event = new GenericEvent();
```
Hmm no. What happens if two events are triggered within the same code paths ? They'd both use the same event object even for different events. Please instantiate `GenericEvent` directly before triggering the event.

sharidas commented:
Instantiation of GenericEvent is moved just before dispatch() is called.

PVince81 commented:
Looks much better. Can you now extend the PHP unit tests to check that the events are fired properly and the values passed in the event object match ?

PVince81 commented:
`make test-php` run in this repo folder should run the tests

sharidas commented:
Updated the PR and added tests to check the event object match GenericEvent. Added tests to check keys in the event hash.
```
sujith@sujith-Inspiron-5567 ~/test/owncloud/apps/customgroups $ make test-php
for F in ./templates/admin.php ./templates/index.php ./tests/integration/features/bootstrap/CustomGroupsContext.php ./tests/unit/CustomGroupsDatabaseHandlerTest.php ./tests/unit/Dav/RolesTest.php ./tests/unit/Dav/MembershipNodeTest.php ./tests/unit/Dav/GroupMembershipCollectionTest.php ./tests/unit/Dav/UsersCollectionTest.php ./tests/unit/Dav/GroupsCollectionTest.php ./tests/unit/Dav/RootCollectionTest.php ./tests/unit/Service/MembershipHelperTest.php ./tests/unit/CustomGroupsPluginTest.php ./tests/unit/CustomGroupsBackendTest.php ./tests/unit/PageControllerTest.php ./tests/unit/SettingsPanelTest.php ./tests/unit/bootstrap.php ./tests/unit/NotifierTest.php ./appinfo/routes.php ./appinfo/app.php ./appinfo/Migrations/Version20161209151129.php ./lib/Notifier.php ./lib/Controller/PageController.php ./lib/Application.php ./lib/CustomGroupsBackend.php ./lib/SettingsSection.php ./lib/CustomGroupsDatabaseHandler.php ./lib/Dav/GroupMembershipCollection.php ./lib/Dav/UsersCollection.php ./lib/Dav/ReportRequest.php ./lib/Dav/GroupsCollection.php ./lib/Dav/MembershipNode.php ./lib/Dav/CustomGroupsPlugin.php ./lib/Dav/RootCollection.php ./lib/Dav/Roles.php ./lib/Service/MembershipHelper.php ./lib/Search.php ./lib/AdminPanel.php ./lib/SettingsPanel.php; do \
        php -l "$F" > /dev/null || exit ; \
done
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
/home/sujith/test/owncloud/apps/customgroups/../../occ app:enable customgroups
Cannot load Xdebug - it was already loaded
customgroups enabled
"/home/sujith/test/owncloud/apps/customgroups"/lib/composer/phpunit/phpunit/phpunit --configuration tests/unit/phpunit.xml --log-junit tests/unit/results/results.xml --coverage-clover tests/unit/results/clover.xml --coverage-html tests/unit/results/coverage-html
Cannot load Xdebug - it was already loaded
PHPUnit 5.7.15 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.0.18-0ubuntu0.16.04.1 with Xdebug 2.4.0
Configuration: /home/sujith/test/owncloud/apps/customgroups/tests/unit/phpunit.xml

...............................................................  63 / 152 ( 41%)
............................................................... 126 / 152 ( 82%)
..........................                                      152 / 152 (100%)

Time: 4.82 seconds, Memory: 32.00MB

OK (152 tests, 490 assertions)

```

PVince81 commented on tests/unit/Dav/GroupsCollectionTest.php:
```diff
@@ -186,6 +187,12 @@ public function testCreateGroup() {
 			->with('user1', 1, true);
 
+		\OC::$server->getEventDispatcher()->addListener('addGroupAndUser', function ($event) {
+			$this->assertTrue($event instanceof GenericEvent);
```
while the test itself might look good, try removing the emitter of the event. The test will likely still pass.
You need to add something that verifies that the handler was really set.

Maybe one way is to add a `use` variable `$called` by reference that you set, then assert outside that it was set.

sharidas replied:
Something that strikes my mind....
```
$called = array();
$dispatcher = \OC::$server->getEventDispatcher();
$dispatcher->addListener('addGroupAndUser', function () use (&$called) {$called[] = 'addGroupAndUser';});
```
and then after line ```$this->collection->createDirectory('group1');``` adding the line
```$this->assertSame(array('addGroupAndUser'), $called);```
This way I can compare event named 'addGroupAndUser' was triggered. But then don't we need to check the value of event hash?

PVince81 replied:
you can also simply store the event object into your `$called` array and check the event's values outside of the function

I don't see any reason why you shouldn't verify that the event values match the correct group name, user name, etc

sharidas commented:
I have updated the PR by updating the unit test with ```use``` and ```$called```. Below is the output of ```make test-php```:

```
sujith@sujith-Inspiron-5567 ~/test/owncloud/apps/customgroups $ make test-php 
for F in ./templates/admin.php ./templates/index.php ./tests/integration/features/bootstrap/CustomGroupsContext.php ./tests/unit/CustomGroupsDatabaseHandlerTest.php ./tests/unit/Dav/RolesTest.php ./tests/unit/Dav/MembershipNodeTest.php ./tests/unit/Dav/GroupMembershipCollectionTest.php ./tests/unit/Dav/UsersCollectionTest.php ./tests/unit/Dav/GroupsCollectionTest.php ./tests/unit/Dav/RootCollectionTest.php ./tests/unit/Service/MembershipHelperTest.php ./tests/unit/CustomGroupsPluginTest.php ./tests/unit/CustomGroupsBackendTest.php ./tests/unit/PageControllerTest.php ./tests/unit/SettingsPanelTest.php ./tests/unit/bootstrap.php ./tests/unit/NotifierTest.php ./appinfo/routes.php ./appinfo/app.php ./appinfo/Migrations/Version20161209151129.php ./lib/Notifier.php ./lib/Controller/PageController.php ./lib/Application.php ./lib/CustomGroupsBackend.php ./lib/SettingsSection.php ./lib/CustomGroupsDatabaseHandler.php ./lib/Dav/GroupMembershipCollection.php ./lib/Dav/UsersCollection.php ./lib/Dav/ReportRequest.php ./lib/Dav/GroupsCollection.php ./lib/Dav/MembershipNode.php ./lib/Dav/CustomGroupsPlugin.php ./lib/Dav/RootCollection.php ./lib/Dav/Roles.php ./lib/Service/MembershipHelper.php ./lib/Search.php ./lib/AdminPanel.php ./lib/SettingsPanel.php; do \
        php -l "$F" > /dev/null || exit ; \
done
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
Cannot load Xdebug - it was already loaded
/home/sujith/test/owncloud/apps/customgroups/../../occ app:enable customgroups
Cannot load Xdebug - it was already loaded
customgroups enabled
"/home/sujith/test/owncloud/apps/customgroups"/lib/composer/phpunit/phpunit/phpunit --configuration tests/unit/phpunit.xml --log-junit tests/unit/results/results.xml --coverage-clover tests/unit/results/clover.xml --coverage-html tests/unit/results/coverage-html
Cannot load Xdebug - it was already loaded
PHPUnit 5.7.15 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.0.18-0ubuntu0.16.04.1 with Xdebug 2.4.0
Configuration: /home/sujith/test/owncloud/apps/customgroups/tests/unit/phpunit.xml

...............................................................  63 / 152 ( 41%)
............................................................... 126 / 152 ( 82%)
..........................                                      152 / 152 (100%)

Time: 6.29 seconds, Memory: 30.00MB

OK (152 tests, 492 assertions)
```
